storage/pebble/execution_data_tracker.go

```diff
@@ -0,0 +1,406 @@
+	if err != nil {
+		return fmt.Errorf("failed to set pruned height value: %w", err)
+	}
+
+	return nil
+}
+
+// Update is used to track new blob CIDs.
+// It can be used to track blobs for both sealed and unsealed
+// heights, and the same blob may be added multiple times for
+// different heights.
+// The same blob may also be added multiple times for the same
+// height, but it will only be tracked once per height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) Update(f storage.UpdateFn) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return f(s.trackBlobs)
+}
+
+// SetFulfilledHeight updates the fulfilled height value,
+// which is the lowest from the highest block heights `h` such that all
+// heights <= `h` are sealed and the sealed execution data
+// has been downloaded or indexed.
+// It is up to the caller to ensure that this is never
+// called with a value lower than the pruned height.
+//
+// No errors are expected during normal operation
+func (s *ExecutionDataTracker) SetFulfilledHeight(height uint64) error {
+	err := operation.UpdateTrackerFulfilledHeight(height)(s.db)
+	if err != nil {
+		return fmt.Errorf("failed to set fulfilled height value: %w", err)
+	}
+
+	return nil
+}
+
+// GetFulfilledHeight returns the current fulfilled height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) GetFulfilledHeight() (uint64, error) {
+	var fulfilledHeight uint64
+
+	err := operation.RetrieveTrackerFulfilledHeight(&fulfilledHeight)(s.db)
+	if err != nil {
+		return 0, err
+	}
+
+	return fulfilledHeight, nil
+}
+
+// setPrunedHeight updates the current pruned height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) setPrunedHeight(height uint64) error {
+	err := operation.UpdateTrackerPrunedHeight(height)(s.db)
+	if err != nil {
+		return fmt.Errorf("failed to set pruned height value: %w", err)
+	}
+
+	return nil
+}
+
+// GetPrunedHeight returns the current pruned height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) GetPrunedHeight() (uint64, error) {
+	var prunedHeight uint64
+
+	err := operation.RetrieveTrackerPrunedHeight(&prunedHeight)(s.db)
+	if err != nil {
+		return 0, err
+	}
+
+	return prunedHeight, nil
+}
+
+// trackBlob tracks a single blob CID at the specified block height.
+// This method first inserts a record of the blob at the specified block height.
+// It then checks if the current block height is greater than the previously recorded
+// latest height for this CID. If so, it updates the latest height to the current block height.
+//
+// Parameters:
+// - blockHeight: The height at which the blob was observed.
+// - c: The CID of the blob to be tracked.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) trackBlob(blockHeight uint64, c cid.Cid) error {
+	err := operation.InsertBlob(blockHeight, c)(s.db)
```

Comment:
and all of these write operations should happen within a single batch

Reply:
[34e04ee](https://github.com/onflow/flow-go/pull/6277/commits/34e04eec57c1bfa33cf840f8b107cc507bcf6c24)